### DISCUSSION TITLE
Supporting dynamic repository configurations

### DISCUSSION THREAD

**Discussion by devversion** [renovatebot/renovate#20867] [Suggest an Idea]
### Type of discussion.

I'm proposing an idea

### Tell us more.

Hi all. I'm just wondering if there was ever a thought in allowing for dynamic configurations. In our projects, the branches are changing based on our releases (e.g. we have a `15.2.x` for the active patch release, or the `main` branch is already the draft branch for a `16.0.0`), and we would like to automatically populate the `baseBranches` field, or have semantic branch-specific rules, like:

```js
// renovate-config.js
const [patchBranch, nextBranch] = await determineActiveReleaseBranches();

export default {
  baseBranches: [nextBranch, patchBranch],
  packageRules: [
    {matchBaseBranches: [patchBranch], followTag: "latest"},
    {matchBaseBranches: [nextBranch],  followTag: "next"},
  ],
}
```

I've tried looking for existing issues, but couldn't find anything. I don't know how feasible it would be, so I thought I would start a discussion first.

**Comment by rarkins** [comment]
`baseBranches` can be a regex pattern. Does that already solve your requirement? If not, please elaborate

**Reply by devversion** [reply]
Thanks for the reply. It's not suitable for our use case unfortunately. Consider we have the following branches:

```js
- main   // active @next version of Angular
- 15.2.x // active @latest version of Angular
- 15.1.x // no longer in active development. Would not like Renovate PRs
- 15.0.x // no longer in active development. Would not like Renovate PRs

// (in practice it can also be a little more nuanced/complex, but this should be a sufficient example).
```

Notice how we have a new branch for every minor. An older minor branch can be replaced by a newer minor branch, at which point the older branches become inactive/or LTS. For those we wouldn't want to have Renovate PRs. I don't think it's possible to have a Regex that will only match e.g. the latest minor branch of a major. i.e. in this case `15.2.x`

**Reply by viceice** [reply]
you can simply set `enabled=false` in branches which shouldn't get updates.

https://docs.renovatebot.com/configuration-options/#usebasebranchconfig

**Reply by rarkins** [reply]
The `config.js` which you're discussing is normally _global_ config, applied for all repos. But have you set it up so that you run against one repo only per config.js?

**Reply by devversion** [reply]
We would need the repository config to be dynamic, since the branches vary based from repository to repository.

`useBaseBranchConfig` is interesting. I think it could work, but what the overall goal would be to avoid having to do manual changes everytime a new "version branch" is created. 

We are using self-hosted, and have 1x instance control 5x repositories, so we cannot compute the base branches there.

**Reply by rarkins** [reply]
The repositories array in config.js can be a complex one (include config per repository) so it's not impossible

**Reply by devversion** [reply]
Interesting. What would be the problem with supporting a repository configuration via `.js`? (if e.g. explicitly allowed in the self-hosted global settings). We are hosting our self-hosted runner in a single place, and with your suggestion in `config.js`, we would no longer be able to keep the configuration local to repository right?

**Reply by rarkins** [reply]
In theory we could support a `renovate.js` file opt-in only for self-hosted

**Reply by devversion** [reply]
That would be pretty cool.

**Reply by rarkins** [reply]
Please create a feature request issue if you'd like
